session/session.go

```diff
@@ -2843,30 +2844,77 @@ func logStmt(execStmt *executor.ExecStmt, s *session) {
 	}
 }
 
-func logGeneralQuery(execStmt *executor.ExecStmt, s *session, isPrepared bool) {
+func genFnGetQuery(execStmt *executor.ExecStmt, s *session, isPrepared bool) func(buf *strings.Builder) string {
 	vars := s.GetSessionVars()
-	if variable.ProcessGeneralLog.Load() && !vars.InRestrictedSQL {
-		var query string
+	return func(buf *strings.Builder) string {
 		if isPrepared {
-			query = execStmt.OriginText()
+			buf.WriteString(execStmt.OriginText())
 		} else {
-			query = execStmt.GetTextToLog()
+			buf.WriteString(execStmt.GetTextToLog())
 		}
 
-		query = executor.QueryReplacer.Replace(query)
+		bufString := buf.String()
+		queryMutable := *(*[]byte)(unsafe.Pointer(&bufString))
+		for i, b := range queryMutable {
+			if b == '\r' || b == '\n' || b == '\t' {
+				queryMutable[i] = ' '
+			}
+		}
 		if !vars.EnableRedactLog {
-			query += vars.PreparedParams.String()
+			buf.WriteString(vars.PreparedParams.String())
 		}
-		logutil.BgLogger().Info("GENERAL_LOG",
-			zap.Uint64("conn", vars.ConnectionID),
-			zap.Stringer("user", vars.User),
-			zap.Int64("schemaVersion", s.GetInfoSchema().SchemaMetaVersion()),
-			zap.Uint64("txnStartTS", vars.TxnCtx.StartTS),
-			zap.Uint64("forUpdateTS", vars.TxnCtx.GetForUpdateTS()),
-			zap.Bool("isReadConsistency", vars.IsIsolation(ast.ReadCommitted)),
-			zap.String("current_db", vars.CurrentDB),
-			zap.String("txn_mode", vars.GetReadableTxnMode()),
-			zap.String("sql", query))
+		return buf.String()
+	}
+}
+
+func genGetUser(user *auth.UserIdentity) func() string {
```

Comment:
It seem that the `e.fnGetUser()` only used in one place. Why we need to pack it as a function var. And not call the `e.GetUser()` instead?

Reply:
@jyz0309 sorry I didn't get it. could you please elaborate that?
btw, what is `e.GetUser()`?

Reply:
I mean that I think we don't need this function to return a function. And I think we don't need the `fnGetUser()` field for `GeneralLogEntry`.

Reply:
This is for breaking the dependency of `logutil` and `session`, otherwise `logutil` must import from `session` which will introduce cyclic dependency.

Reply:
removed.